File: src/main/java/de/bonndan/nivio/output/map/NLabel.java
package de.bonndan.nivio.output.map;

import j2html.tags.DomContent;

import static j2html.TagCreator.div;

class NLabel extends Component {

    private final ItemMapItem item;
    private final String className;
    final int width, size, padding;

    NLabel(ItemMapItem item, String className, int width, int size, int padding) {
        this.item = item;
        this.className = className;
        this.width = width;
        this.size = size;
        this.padding = padding;
    }


    public DomContent render() {
        var style = "stroke: " + item.status;

        /*
        {item.landscapeItem.description && (<div>"{item.landscapeItem.description}"<br/></div>)}
        {item.landscapeItem.owner && (<div>Owner: {item.landscapeItem.owner}</div>)}
        {item.landscapeItem.team && (<div>Team: {item.landscapeItem.team}</div>)}
        {item.landscapeItem.contact && (<div>Contact: {item.landscapeItem.contact}</div>)}
        {item.landscapeItem.software && (<div>Software: {item.landscapeItem.software}</div>)}
        {item.landscapeItem.version && (<div>Version: {item.landscapeItem.version}</div>)}
        {item.landscapeItem.lifecycle && (<div>Lifecycle: {item.landscapeItem.lifecycle}</div>)}
         */

        var rect = SvgTagCreator.rect()
                .attr("x", size + padding)
                .attr("y", -10)
                .attr("rx", 10)
                .attr("ry", 10)
                .attr("fill", "white")
                .attr("width", width)
                .attr("height", size / 2)
                .attr("style", style);
        NText nText = new NText(item, size + padding + (width / 2), 5, "", width);
        var fo = SvgTagCreator.foreignObject(div("").attr("class", "details"))
                .attr("width", width)
                .attr("height", 220)
                .attr("y", padding)
                .attr("x", size + padding);

        return SvgTagCreator.g(rect, nText.render(), fo).attr("class", "label");
    }
}

File: src/main/app/src/hexmap/NLabel.js
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import NText from './NText';

class NLabel extends Component {
  static propTypes = {
    item: PropTypes.object,
    x: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    y: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    className: PropTypes.string,
    width: PropTypes.number,
    size: PropTypes.number,
    padding: PropTypes.number,
  };

  constructor() {
    super();
    this.state = { showDetails: false };
  }

  itemPopup(e) {
    this.setState({ showDetails: !this.state.showDetails });
  }

  render() {
    const { size, item, padding, width } = this.props;
    let showDetails = this.state.showDetails;
    let style = {
      stroke: item.status,
    };
    let detailsStyle = {
      display: showDetails ? 'block' : 'none',
    };

    return (
      <g className={'label'}>
        <rect
          x={size + padding}
          y={-10}
          rx='10'
          ry='10'
          fill={'white'}
          width={width}
          height={size / 2}
          style={style}
        />
        <NText
          key={item.identifier}
          x={size + padding + width / 2}
          y={5}
          width={width}
          item={item}
          onClick={e => this.itemPopup(e, this)}
        />

        <foreignObject
          width={width}
          height='220'
          y={padding}
          x={size + padding}
          style={detailsStyle}
        >
          <div className='details'>
            {item.landscapeItem.description && (
              <div>
                "{item.landscapeItem.description}"<br />
              </div>
            )}
            {item.landscapeItem.owner && <div>Owner: {item.landscapeItem.owner}</div>}
            {item.landscapeItem.team && <div>Team: {item.landscapeItem.team}</div>}
            {item.landscapeItem.contact && <div>Contact: {item.landscapeItem.contact}</div>}
            {item.landscapeItem.software && <div>Software: {item.landscapeItem.software}</div>}
            {item.landscapeItem.version && <div>Version: {item.landscapeItem.version}</div>}
            {item.landscapeItem.lifecycle && <div>Lifecycle: {item.landscapeItem.lifecycle}</div>}
          </div>
        </foreignObject>
      </g>
    );
  }
}

export default NLabel;
